File: src/main/java/io/wordii/nlp/Sentence.java
package io.wordii.nlp;

import com.google.common.graph.ValueGraph;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class Sentence {
    private static final Pattern PATTERN = Pattern.compile("\".+\"");

    @NotNull
    private final String rawText;
    @NotNull
    private final ValueGraph<Word, Relation> dependencyTree;
    @NotNull
    private final List<Word> roots;
    // Note: words and punctuations (aka tokens)
    @NotNull
    private final List<Word> wordList;
    // Position in text
    private final int position;

    public Sentence(@NotNull String rawText, @NotNull ValueGraph<Word, Relation> dependencyTree, @NotNull List<Word> roots, @NotNull List<Word> wordList, int position) {
        this.rawText = rawText;
        this.dependencyTree = dependencyTree;
        this.roots = roots;
        this.wordList = wordList;
        this.position = position;
    }

    @NotNull
    public Set<Word> children(@NotNull Word node) {
        return dependencyTree.successors(node);
    }

    @Nullable
    public Word childWithEdgeValue(@NotNull Word node, @NotNull Relation edgeValue) {
        Set<Word> successors = dependencyTree.successors(node);
        for (Word successor : successors) {
            if (edgeValue.equals(edgeValue(node, successor))) {
                return successor;
            }
        }
        return null;
    }

    @Nullable
    public Relation edgeValue(@NotNull Word nodeU, @NotNull Word nodeV) {
        return dependencyTree.edgeValue(nodeU, nodeV).orElse(null);
    }

    public boolean hasDirectSpeech() {
        return PATTERN.matcher(rawText).find();
    }

    @NotNull
    public List<String> getDirectSpeeches() {
        if (!hasDirectSpeech()) {
            return Collections.emptyList();
        }
        Matcher matcher = PATTERN.matcher(rawText);
        List<String> speeches = new ArrayList<>();
        while (matcher.find()) {
            String group = matcher.group();
            speeches.add(group.substring(1, group.length() - 1));
        }
        return Collections.unmodifiableList(speeches);
    }

    @NotNull
    public String getRawText() {
        return rawText;
    }

    @NotNull
    public List<Word> getRoots() {
        return roots;
    }

    @NotNull
    public List<Word> getWordList() {
        return wordList;
    }

    public int getPosition() {
        return position;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Sentence sentence = (Sentence) o;
        return rawText.equals(sentence.rawText);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rawText);
    }

    @Override
    public String toString() {
        return "Sentence{" +
                "rawText='" + rawText + '\'' +
                ", dependencyTree=" + dependencyTree +
                ", roots=" + roots +
                ", wordList=" + wordList +
                ", position=" + position +
                '}';
    }
}

File: src/main/kotlin/io/wordii/nlp/Sentence.kt
package io.wordii.nlp

import com.google.common.graph.ValueGraph
import kotlinx.collections.immutable.toImmutableList
import java.util.regex.Pattern

data class Sentence(
    val rawText: String,
    private val dependencyTree: ValueGraph<Word, Relation>,
    val roots: List<Word>, // Note: words and punctuations (aka tokens)
    val wordList: List<Word>, // Position in text
    val position: Int
) {
    fun children(node: Word): Set<Word> {
        return dependencyTree.successors(node)
    }

    fun childWithEdgeValue(node: Word, edgeValue: Relation): Word? {
        val successors = dependencyTree.successors(node)
        for (successor in successors) {
            if (edgeValue == edgeValue(node, successor)) {
                return successor
            }
        }
        return null
    }

    fun edgeValue(nodeU: Word, nodeV: Word): Relation? {
        return dependencyTree.edgeValue(nodeU, nodeV).orElse(null)
    }

    fun hasDirectSpeech(): Boolean {
        return PATTERN.matcher(rawText).find()
    }

    val directSpeeches: List<String>
        get() {
            if (!hasDirectSpeech()) {
                return emptyList()
            }
            val matcher = PATTERN.matcher(rawText)
            val speeches = mutableListOf<String>()
            while (matcher.find()) {
                val group = matcher.group()
                speeches.add(group.substring(1, group.length - 1))
            }
            return speeches.toImmutableList()
        }

    companion object {
        private val PATTERN = Pattern.compile("\".+\"")
    }
}
